Bazel query parsing with protobuf

sean7218 opened the pull request:
## ✨ Description

Add protobuf streaming support for Bazel query operations
This PR introduces support for parsing Bazel query results using protobuf streaming format (--output streamed_proto) as an alternative to the existing XML-based approach. This improvement provides better performance and more structured data handling for dependency queries.

##  🚀  Key Changes

- Added `queryTargetsWithProto()` method to `BazelTargetQuerier` that uses `--output streamed_proto` instead of `XML`
- Implemented `parseQueryTargets()` in `BazelProtobufBindings` to parse streamed protobuf messages
- Added `parseTargetsWithProto()` method to `BazelQueryParser` for handling `BlazeQuery_Target` protobuf objects
- Removed `--output xml` query and parsing

## 🧪 Testing & Validation:

- Added comprehensive test suite with `BazelTargetParserTests` for protobuf functionality
- Updated `BazelTargetQuerierTests` with protobuf-specific test cases
- Added test resources `streamdeps.pb` for validation of protobuf parsing
- Enhanced `BazelProtobufBindingsTests` with streaming protobuf decode tests

sean7218 committed:
Add build proto and stardoc_output proto definitions and update tests

- Introduced a new proto file `build.proto` containing definitions for target data, including target name, source files, and rules.
- Introduced a new proto file `stardoc_output.proto` containing definitions for Stardoc data, including messages for module information, rule definitions, provider information, and function definitions.
- Updated `BazelProtobufBindingsTests.swift` to replace the deprecated action parser with the new `parseActionGraph` method and added a new test for decoding stream proto data.
- Added a new binary resource file `streamdeps.pb` for testing purposes.

sean7218 committed:
Merge remote-tracking branch 'origin/main' into sean7218/proto-streamdeps

sean7218 committed:
update BazelProtobufBindingsTests tests and add swift-cmark to gitignore

run swift-format

sean7218 committed:
Add BazelProtobufBindings integration and feature flag for protobuf parsing

sean7218 committed:
fix the query issue with Data encoding with .utf8

sean7218 committed:
added query parser to proto

sean7218 committed:
instead of throw, just log src file not found

sean7218 committed:
fix the file scheme for the src file

sean7218 committed:
add the testing file streamdeps.pb

sean7218 commented on Sources/SourceKitBazelBSP/RequestHandlers/BuildTargets/BazelTargetQuerier.swift:
```diff
@@ -98,3 +99,75 @@ final class BazelTargetQuerier {
+}
+
+extension CommandRunner {
+    func getStdout(with cmd: String, cwd: String?) throws -> Data {
```
need to refactor this code with CommandRunner protocol

sean7218 commented on Sources/SourceKitBazelBSP/Server/InitializedServerConfig.swift:
```diff
@@ -37,4 +37,12 @@ struct InitializedServerConfig: Equatable {
+    /// when dislabled, the default bazel query is parsed in xml format
+    var protoMode: Bool {
+        // return ProcessInfo.processInfo.environment["PROTO_MODE"] == "1"
+        // TODO: toggle back
```
TODO: toggle this back when PR is ready for review

sean7218 committed:
Add streamdeps resource and enhance command runner functionality

- Added `streamdeps.pb` resource to the test target.
- Updated `BazelQueryParser` tests to utilize the new resource.
- Refactored `CommandRunner` protocol and its implementations for better command execution handling.
- Introduced `CommandRunnerFake` to support data responses in tests.

sean7218 committed:
Fix linter issues

rockbruno commented on Sources/SourceKitBazelBSP/SharedUtils/Shell/CommandRunner.swift:
```diff
@@ -20,11 +20,19 @@
 import Foundation
 
 protocol CommandRunner {
+    func execute(_ cmd: String, cwd: String?) throws -> Data
```
Nice one! A small nit: I think we can actually keep the same name `run` name here, then each caller chooses if it wants the string or the raw data

sean7218 committed:
Refactor BazelProtobufBindings from struct to enum and simplify testOnly attribute extraction in BazelQueryParser

sean7218 committed:
Refactor Bazel query handling to use protobuf format and streamline target parsing

sean7218 committed:
fixed tests when changing from xml to proto parsing

sean7218 committed:
Simplify call sites for CommandRunner

- remove execute method from CommandRunner
- refactored the output from String to Data for run(cmd:cwd) -> String
- added toStr() for conversion from Data to String

sean7218 commented on Sources/BazelProtobufBindings/BazelProtobufBindings.swift:
```diff
@@ -19,20 +19,115 @@
+extension BazelProtobufBindings {
+    /// Bazel query outputs a series of messages and each one is prefixed with length to indcate
+    /// number of bytes in the payload. Returns a tuple of (value, bytesConsumed)
+    static func parseVarint(
```
added it to the comment section 
https://protobuf.dev/programming-guides/encoding/

sean7218 commented on Sources/SourceKitBazelBSP/RequestHandlers/BuildTargets/BazelQueryParser.swift:
```diff
@@ -17,114 +17,196 @@
+            }.compactMap { src in
+                guard let path = srcMap[src] else {
+                    let error = BazelTargetParserError.noSrcFound(src)
+                    logger.info("\(error, privacy: .public)")
```
**qq**: The genrule files are not yet generated at this point of time, do we need to handle these generated files somehow? Because we need to mark the generated files for the `TargetSourceHandler`

cc: @rockbruno @josh-arnold-1

sean7218 committed:
fix linter issue

rockbruno commented on Sources/SourceKitBazelBSP/RequestHandlers/BuildTargets/BazelQueryParser.swift:
```diff
@@ -17,114 +17,196 @@
+                return try URI(string: path)
             }
+
+            let data = SourceKitBuildTarget(
+                toolchain: try URI(string: "file://" + toolchainPath)
+            ).encodeToLSPAny()
```
Since this is the same for all libraries, we can move it outside of the loop for a small perf gain.

rockbruno commented on Sources/SourceKitBazelBSP/RequestHandlers/BuildTargets/BazelQueryParser.swift:
```diff
@@ -17,114 +17,196 @@
+    /// Bazel query outputs a list of targets and each target contains list of attributes.
+    /// The `srcs` attribute is a list of source_file labels instead of URI, thus we need
+    /// a hashmap to reduce the time complexity.
+    static func buildSourceFilesMap(
```
The previous logic had some additional checks that are not present here (particularly filtering against only local content `//` since I hadn't played with third party repos at the time), I think we might still need those, but am not 100% sure. I guess as long as it works for you and @josh-arnold-1 then it should be fine.

rockbruno commented on Sources/SourceKitBazelBSP/RequestHandlers/BuildTargets/BazelQueryParser.swift:
```diff
@@ -17,114 +17,196 @@
+// MARK: - Bazel target name helpers
+
+extension String {
+    /// Convert the target name into file URI scheme and returns unique target id
```
```suggestion
    /// Converts the target name into a URI and returns a unique target id.
```

rockbruno commented on Sources/SourceKitBazelBSP/RequestHandlers/BuildTargets/BazelQueryParser.swift:
```diff
@@ -17,114 +17,196 @@
+        return uri
+    }
+
+    /// Convert the target name into file URI scheme and returns the directory
```
```suggestion
    /// Converts the target name a URI and returns the target's base directory.
```

rockbruno commented on Sources/SourceKitBazelBSP/RequestHandlers/BuildTargets/BazelQueryParser.swift:
```diff
@@ -17,114 +17,196 @@
 
-    static func bazelTargetToURI(_ bazelTarget: String) -> String {
-        return "file://\(bazelTarget.replacingOccurrences(of: ":", with: "___"))"
+    /// Split bazel target label into packageName and targetName
```
```suggestion
    /// Splits a full Bazel label into a tuple of its package and target names.
```

rockbruno commented on Sources/SourceKitBazelBSP/RequestHandlers/BuildTargets/BazelTargetQuerier.swift:
```diff
@@ -59,7 +60,12 @@ final class BazelTargetQuerier {
     }
 
-    func queryTargets(forConfig config: BaseServerConfig, rootUri: String, kinds: Set<String>) throws -> XMLElement {
+    /// Based on the server config, it constructs query and calls `bazel query deps` with output proto
+    func queryTargets(
```
```suggestion
    func queryTargets(
```

Ok to not have the comment in this case

rockbruno commented on Sources/SourceKitBazelBSP/SharedUtils/Shell/CommandRunner.swift:
```diff
@@ -20,18 +20,27 @@
+    func run(_ cmd: String, cwd: String?) throws -> Data
+}
+
+extension Data {
+    func toStr() -> String {
+        let str = String(data: self, encoding: .utf8) ?? ""
+        return str.trimmingCharacters(in: .whitespacesAndNewlines)
+    }
 }
```
```suggestion
extension CommandRunner {
    func run(_ cmd: String, cwd: String?) throws -> String {
        let data: Data = try run(cmd, cwd: cwd)
        let str = String(data: data, encoding: .utf8) ?? ""
        return str.trimmingCharacters(in: .whitespacesAndNewlines)
    }

    func run(_ cmd: String) throws -> String {
        try run(cmd, cwd: nil)
    }
}
```

Since this is also trimming lines, I think it will be more accurate to keep the call under CommandRunner itself like it was before. You can then disambiguate by adding the type directly to the callers (`: String`).

rockbruno commented on Tests/BazelProtobufBindingsTests/BazelProtobufBindingsTests.swift:
```diff
@@ -108,4 +107,31 @@ struct BazelProtobufBindingsTests {
         #expect(actual == expected)
     }
+
+    @Test
+    func testDecode_streamProto() throws {
```
```suggestion
    @Test
    func decodesStreamProto() throws {
```

No more need for the test prefix when using Swift Testing nowadays

sean7218 committed:
Merge branch 'main' of github.com:sean7218/sourcekit-bazel-bsp into sean7218/proto-streamdeps

sean7218 committed:
Refactor methods to improve clarity and encapsulation; update test function naming for consistency

maxwellE commented on Sources/BazelProtobufBindings/protos/build.proto:
Conceptually I am wondering why we need to commit these or the pb.swift files, shouldn't this just be provided by Bazel in the sandbox instead of being committed to the repo? 

Unsure of the value of appeasing SPMs limitations here when people using this will be using bazel

sean7218 committed:
Add buildTestSuffix parameter to Bazel target parsing and tests

sean7218 commented on Sources/SourceKitBazelBSP/RequestHandlers/BuildTargets/BazelQueryParser.swift:
```diff
@@ -17,114 +17,196 @@
+            // get test_only
+            let testOnly = rule.attribute.first { $0.name == "testonly" }?.booleanValue ?? false
 
-        // FIXME: Most of this logic is hacked together and not thought through, with the
-        // sole intention of getting the example project to work.
-        // Need to understand what exactly we can receive from the queries to know how to properly
-        // parse this info.
```
added below

sean7218 committed:
Fix indentation in parseVarint documentation comment
